File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Unit/Monster/ViewModel/TheBottom/GremlinFat_ViewModel.ts
import {AbstractMonster} from "mods/ModTheSpire/Scripts/Unit/Monster/AbstractMonster";
import DamageInfo from "mods/ModTheSpire/Scripts/DataDefine/DamageInfo";
import {Mathf} from "Core/Module/Math/Mathf";
import {Spine} from "csharp";
import DungeonManager from "mods/ModTheSpire/Scripts/DungeonManager";
import {LocalizationProperty} from "mods/ModTheSpire/Scripts/Gen/DB/Localization";
import {ApplyPowerAction} from "mods/ModTheSpire/Scripts/Action/Common/ApplyPowerAction";
import DamageAction from "mods/ModTheSpire/Scripts/Action/Common/DamageAction";
import {AttackEffect} from "mods/ModTheSpire/Scripts/DataDefine/AttackEffect";
import {RollMoveAction} from "mods/ModTheSpire/Scripts/Action/Common/RollMoveAction";
import {Intent} from "mods/ModTheSpire/Scripts/Unit/Monster/Intent";
import {DamageType} from "mods/ModTheSpire/Scripts/DataDefine/DamageType";
import {SFXAction} from "mods/ModTheSpire/Scripts/Action/Utility/SFXAction";
import {AnimateSlowAttackAction} from "mods/ModTheSpire/Scripts/Action/Animations/AnimateSlowAttackAction";
import {WeakPower} from "mods/ModTheSpire/Scripts/Power/WeakPower";
import {FrailPower} from "mods/ModTheSpire/Scripts/Power/FrailPower";
import {SetMoveAction} from "mods/ModTheSpire/Scripts/Action/Common/SetMoveAction";
import {SpeechBubble} from "mods/ModTheSpire/Scripts/Effect/Combat/SpeechBubble";
import {EscapeAction} from "mods/ModTheSpire/Scripts/Action/Common/EscapeAction";

export class GremlinFat_ViewModel extends AbstractMonster{
    Initialize() {
        if(DungeonManager.Inst.AdvanceLevel >= 2) {
            this.DamageInfo.push(new DamageInfo(this,5));
        }else{
            this.DamageInfo.push(new DamageInfo(this,4));
        }
        let trackEntry = <Spine.TrackEntry>this.SetAnimation("animation",true);
        trackEntry.TrackTime = trackEntry.TrackEnd * Mathf.Random()
    }

    GetMove(num: number) {
        this.SetMove(2,Intent.ATTACK_DEBUFF,{moveName:LocalizationProperty.Read(this.Info.Name + "行动1")})
    }

    TakeTurn() {
        switch(this.MoveIndex){
            case 2:{
                this.AddToBot(new AnimateSlowAttackAction(this));
                this.AddToBot(new DamageAction(DungeonManager.Inst.Player,this.DamageInfo[0],AttackEffect.BLUNT_LIGHT));
                this.AddToBot(new ApplyPowerAction(DungeonManager.Inst.Player,this,new WeakPower(),1));
                if(DungeonManager.Inst.AdvanceLevel >= 17){
                    this.AddToBot(new ApplyPowerAction(DungeonManager.Inst.Player,this,new FrailPower(),1));
                }
                this.AddToBot(new RollMoveAction(this));
                break;
            }
            case 99:{
                this.PlaySfx()
                DungeonManager.EffectManager.Play(new SpeechBubble(this.X - 100,this.Y + 50,LocalizationProperty.Read(this.Info.Name + "对话1"),2.5,true));
                this.AddToBot(new EscapeAction(this));
                this.AddToBot(new SetMoveAction(this,99,Intent.ESCAPE));
                break;
            }
        }
    }

    private PlaySfx(){
        let roll = Mathf.Floor(Mathf.RandomRange(0,3));
        if(roll == 0){
            this.AddToBot(new SFXAction("VO_GREMLINFAT_1A"));
        } else if(roll == 1){
            this.AddToBot(new SFXAction("VO_GREMLINFAT_1B"));
        } else {
            this.AddToBot(new SFXAction("VO_GREMLINFAT_1C"));
        }
    }

    private PlayDeathSfx(){
        let roll = Mathf.Floor(Mathf.RandomRange(0,3));
        if(roll == 0){
            this.AddToBot(new SFXAction("VO_GREMLINFAT_2A"));
        } else if(roll == 1){
            this.AddToBot(new SFXAction("VO_GREMLINFAT_2B"));
        } else {
            this.AddToBot(new SFXAction("VO_GREMLINFAT_2C"));
        }
    }

    Die() {
        super.Die();
        this.PlayDeathSfx();
    }
}

File: Unity/Assets/Scripts/ModTheSpire/Unit/Monster/ViewModel/TheBottom/GremlinFat_ViewModel.cs
using ModTheSpire.DataDefine;
using ModTheSpire.Dungeon;
using ModTheSpire.Effect;
using ModTheSpire.Effect.Combat;
using ModTheSpire.Effect.Other;
using ModTheSpire.Power.Debuff;
using Spine;
using Spine.Unity;
using TCG.Config;
using UnityEngine;

namespace ModTheSpire.Unit.Monster.ViewModel.TheBottom
{
    public class GremlinFat_ViewModel : AbstractMonster
    {
        protected override void Initialize()
        {
            DamageInfo.Add(new DamageInfo(this, DungeonManager.Inst.AdvanceLevel >= 2 ? 5 : 4));
            var trackEntry = (TrackEntry)this.DisplayObject.SetAnimation("animation", true);
            trackEntry.TrackTime = trackEntry.TrackEnd * Random.value;
        }

        protected override void GetMove(int num)
        {
            this.SetMove(2, Intent.ATTACK_DEBUFF, moveName: LocalizationProperty.Read(this.Info.Name + "行动1"));
        }

        protected override void TakeTurn()
        {
            switch(this.MoveIndex){
                case 2:{
                    this.AddToBot(new AnimateSlowAttackAction(this));
                    this.AddToBot(new DamageAction(DungeonManager.Inst.Player,this.DamageInfo[0],AttackEffect.BLUNT_LIGHT));
                    this.AddToBot(new ApplyPowerAction(DungeonManager.Inst.Player,this,new WeakPower(),1));
                    if(DungeonManager.Inst.AdvanceLevel >= 17){
                        this.AddToBot(new ApplyPowerAction(DungeonManager.Inst.Player,this,new FrailPower(),1));
                    }
                    this.AddToBot(new RollMoveAction(this));
                    break;
                }
                case 99:
                {
                    this.PlaySfx();
                    DungeonManager.EffectManager.Play(new SpeechBubble(this.DisplayObject.X - 100, this.DisplayObject.Y + 50,
                        LocalizationProperty.Read(this.Info.Name + "对话1"), 2500, true));
                    this.AddToBot(new EscapeAction(this));
                    SetMove(99,Intent.ESCAPE);
                    break;
                }
            }
        }

        public override void Damage(DamageInfo info)
        {
            base.Damage(info);
            if(info.Type == DamageType.THORNS && info.Total > 0){
                var animation = (SkeletonAnimation)this.DisplayObject.Self;
                animation.AnimationState.SetAnimation(0,"Hit",false);
                animation.AnimationState.AddAnimation(0,"Idle",true,0);
            }
        }
        
        private void PlaySfx()
        {
            int roll = Random.Range(0, 3);
            DungeonManager.ActionManager.AddToBottom(new SFXAction(roll switch
            {
                0 => SoundMaster.VO_GREMLINFAT_1A,
                1 => SoundMaster.VO_GREMLINFAT_1B,
                _ => SoundMaster.VO_GREMLINFAT_1C,
            }));
        }
        
        private void PlayDeathSfx()
        {
            int roll = Random.Range(0, 3);
            DungeonManager.ActionManager.AddToBottom(new SFXAction(roll switch
            {
                0 => SoundMaster.VO_GREMLINFAT_2A,
                1 => SoundMaster.VO_GREMLINFAT_2B,
                _ => SoundMaster.VO_GREMLINFAT_2C,
            }));
        }

        protected override void Die()
        {
            base.Die();
            PlayDeathSfx();
        }

        public GremlinFat_ViewModel(MonsterInfo info) : base(info)
        {
        }
    }
}